From 0f218e07bf3d6ec9744d203ba3a774baa7ca1c0a Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Wed, 10 Sep 2014 18:28:50 -0400
Subject: [ethernet] ixgbe: Don't receive packets when the napi budget == 0

Message-id: <1410373750-11341-31-git-send-email-jogreene@redhat.com>
Patchwork-id: 93534
O-Subject: [RHEL7.1 PATCH 30/50] ixgbe: Don't receive packets when the napi budget == 0
Bugzilla: 1091123
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ: 1091123
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

Processing any incoming packets with a with a napi budget of 0
is incorrect driver behavior.

This matters as netpoll will shortly call drivers with a budget of 0
to avoid receive packet processing happening in hard irq context.

Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit fdabfc8a74c713f4e4318715d449651f798db74a)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c b/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
index 1beeea1..e1d532c 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
@@ -2078,7 +2078,7 @@ static int ixgbe_clean_rx_irq(struct ixgbe_q_vector *q_vector,
 #endif /* IXGBE_FCOE */
  u16 cleaned_count = ixgbe_desc_unused(rx_ring);
 
- do {
+ while (likely(total_rx_packets < budget)) {
   union ixgbe_adv_rx_desc *rx_desc;
   struct sk_buff *skb;
 
@@ -2153,7 +2153,7 @@ static int ixgbe_clean_rx_irq(struct ixgbe_q_vector *q_vector,
 
   /* update budget accounting */
   total_rx_packets++;
- } while (likely(total_rx_packets < budget));
+ }
 
  u64_stats_update_begin(&rx_ring->syncp);
  rx_ring->stats.packets += total_rx_packets;
-- 
1.7.1